Add FeatureDetector caching.

StoneyJackson opened the pull request:
Implements #36

StoneyJackson committed:
Add FeatureDetector caching.

kevin-brown commented on src/mousetrap/vision.py:
```diff
@@ -87,6 +88,23 @@ class HaarNameError(Exception):
+
+    @classmethod
+    def get_detector(cls, config, name, scale_factor=1.1, min_neighbors=3):
+        if name in cls._INSTANCES:
```
Considering the name is only unique to the haar file used and not the arguments (such as the scale factor or minimum neighbours), we should be storing the combination based on a different key.

StoneyJackson replied:
Good point.

StoneyJackson committed:
Adjust key used to cache FeatureDetectors.

Considering the name is only unique to the haar file used and not
the arguments (such as the scale factor or minimum neighbours),
we should be storing the combination based on a different key.

StoneyJackson commented:
Are we good to merge?

kevin-brown commented on src/mousetrap/vision.py:
```diff
@@ -106,15 +126,28 @@ def __init__(self, config, name, scale_factor=1.1, min_neighbors=3):
+                    {'image':id(image), 'result':self._detect_cache[image]}
+                    )
+            if isinstance(self._detect_cache[image], FeatureNotFoundException):
+                raise FeatureNotFoundException(str(self._detect_cache[image]))
```
Is it possible to re-raise the stored exception? While manually re-raising the new exception while using the old `message` works, we will lose any extra data that is stored within the exception (such as the stack trace) and it locks the implementation down quite a bit.

kevin-brown commented on src/mousetrap/image.py:
```diff
@@ -36,6 +36,12 @@ def get_width(self):
+        return id(self)
+
+    def __eq__(self, other):
+        return hash(self) == hash(other)
```
What required us to re-implement both this and the object hash?

StoneyJackson committed:
Remove supurfluous methods.

StoneyJackson committed:
Chain exceptions for FeatureNotFoundException on cache hits.

Clients can now retreive the original exception that caused the
new exception as follows

```
try:
    detector.detect(image)
except FeatureNotFoundException as exception:
    print exception    # message contains information about cause too.
    if exception.cause is not None:
        print exception.cause
```

StoneyJackson committed:
Call parent constructor in FeatureNotFoundException.
